### PR TITLE
ROX-13164 compliance to retry sending node inventory

### PR DESCRIPTION
**Pull request by jschnath**
## Description

Adds communication between Compliance and Sensor. Sensor sends a NACK message to Compliance if it receives a NodeInventory while it's not ready yet. Compliance upon receiving the NACK resends the NodeInventory after a timeout.
//TODO(Update the description)

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~Evaluated and added CHANGELOG entry if required~
- [ ] ~Determined and documented upgrade steps~
- [ ] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

Manual testing performed, Ack received as expected. We forced sensor to send a dummy message to compliance on every single occurence of NodeInventory and we observed the logs of sensor and compliance to confirm that the Ack was sent as intended.

### CODE REVIEW

File: compliance/collection/main.go
Comment (jschnath) [inline compliance/collection/main.go:318]: fromSensorC has been deleted, remove it from this comment

File: compliance/collection/main.go
Comment (jschnath) [inline compliance/collection/main.go:327]: No need to multiplex here

File: pkg/channelmultiplexer/channelmultiplexer.go
Comment (jschnath) [inline pkg/channelmultiplexer/channelmultiplexer.go:15]: remove these comments

File: pkg/channelmultiplexer/channelmultiplexer.go
Comment (jschnath) [inline pkg/channelmultiplexer/channelmultiplexer.go:35]: Reword this nicer

File: pkg/channelmultiplexer/channelmultiplexer_test.go
Comment (jschnath) [inline pkg/channelmultiplexer/channelmultiplexer_test.go:16]: check if this really needs to be chan *string, otherwise make it chan string

Comment (jschnath) [reply pkg/channelmultiplexer/channelmultiplexer_test.go:16]: Make 3 channels before the test and close them afterwards

File: proto/internalapi/sensor/compliance_iservice.proto
Comment (jschnath) [inline proto/internalapi/sensor/compliance_iservice.proto:53]: Add a comment here that the Nack will be removed and that's why its not consolidated with the Ack

File: sensor/common/compliance/multiplexer.go
Comment (jschnath) [inline sensor/common/compliance/multiplexer.go:33]: Remove this debug message

File: sensor/common/compliance/node_inventory_handler_impl.go
Comment (jschnath) [inline sensor/common/compliance/node_inventory_handler_impl.go:86]: package this routine into a separate function

File: sensor/common/compliance/node_inventory_handler_impl.go
Comment (jschnath) [inline sensor/common/compliance/node_inventory_handler_impl.go:114]: Make this a debug message

File: sensor/common/compliance/service_impl.go
Comment (jschnath) [inline sensor/common/compliance/service_impl.go:174]: move this back to where it was to keep the diff smaller

File: sensor/common/component.go
Comment (jschnath) [inline sensor/common/component.go:38]: ComplianceComponent is a sensor component that can communicate with compliance. All the messaged intended for compliance are returned by ComplianceC(). It _must_ be started before the compliance.Multiplexer or we panic.

File: sensor/common/sensor/sensor.go
Comment (jschnath) [inline sensor/common/sensor/sensor.go:214]: double check that there is no duplicate of this message

File: sensor/kubernetes/sensor/sensor.go
Comment (jschnath) [inline sensor/kubernetes/sensor/sensor.go:102]: Add to the followup ticket that auditLogEventsInput and auditLogCollectionManager should turn into ComplianceComponent if sensible

File: proto/internalapi/sensor/compliance_iservice.proto
Comment (Maddosaurus) [reply proto/internalapi/sensor/compliance_iservice.proto:53]: Should we have this as a //TODO(ROX-xxxx) item with a follow up so we don't forget about it?

**Comment by vikin91** [comment]
Any idea why the `Unit Tests / go (GOTAGS=release)` fail?

File: sensor/common/compliance/multiplexer.go
Comment (Maddosaurus) [reply sensor/common/compliance/multiplexer.go:33]: Is this still open or is there maybe a push missing?

File: pkg/channelmultiplexer/channelmultiplexer_test.go
Comment (vikin91) [reply pkg/channelmultiplexer/channelmultiplexer_test.go:16]: Oh, shall we then change the template to be `<-chan T` instead of `<-chan *T`?

File: sensor/common/compliance/node_inventory_handler.go
Comment (vikin91) [inline sensor/common/compliance/node_inventory_handler.go:13]: Why we cannot return `*nodeInventoryHandlerImpl` here?

File: sensor/common/component.go
Comment (Maddosaurus) [inline sensor/common/component.go:38]:
```suggestion
// ComplianceComponent is a sensor component that can communicate with compliance. All the messages intended for
```

File: sensor/common/compliance/node_inventory_handler_impl.go
Comment (vikin91) [inline sensor/common/compliance/node_inventory_handler_impl.go:115]: Maybe we could stick to the message fragment that we used before - i.e., `Requesting Compliance to resend NodeInventory later` and then `after receiving unknown NodeInventory with ID %s`?

File: sensor/common/compliance/node_inventory_handler_impl.go
Comment (vikin91) [inline sensor/common/compliance/node_inventory_handler_impl.go:123]: This must be removed before merging - we used that only for testing the communication

**Review by Maddosaurus — APPROVED** [review]
Some small thoughts I have, but overall LGTM!

File: sensor/common/compliance/node_inventory_handler_impl.go
Comment (vikin91) [inline sensor/common/compliance/node_inventory_handler_impl.go:160]: This func may be removed as well when we remove the call to it in line 123.

File: sensor/common/component.go
Comment (vikin91) [inline sensor/common/component.go:38]:
```suggestion
// ComplianceComponent is a sensor component that can communicate with compliance. All messages intended for
```

File: pkg/channelmultiplexer/channelmultiplexer_test.go
Comment (vikin91) [reply pkg/channelmultiplexer/channelmultiplexer_test.go:16]: It looks like we forgot the "Make 3 channels before the test and close them afterwards"

**Review by vikin91 — CHANGES_REQUESTED** [review]
I gave it another look and we forgot to remove sending ack - it was there to test that feat and I think it should not be merged.

Moreover, I would like to try using `<-chan T` instead of `<-chan *T` because we do not want to force developer using pointers if that is not necessary.

File: pkg/channelmultiplexer/channelmultiplexer_test.go
Comment (jschnath) [reply pkg/channelmultiplexer/channelmultiplexer_test.go]: With 3 channels I got race conditions and the test got stuck. Repeated usage of a channel between test cases I think led to the multiplexer of the first test sucking up the message for the second test. It works with 5 channels.

File: sensor/common/compliance/multiplexer.go
Comment (jschnath) [reply sensor/common/compliance/multiplexer.go]: Must've overlooked this one, yeah. I removed it now.

File: sensor/common/sensor/sensor.go
Comment (jschnath) [reply sensor/common/sensor/sensor.go:214]: There are no duplicates 👍

File: sensor/common/component.go
Comment (jschnath) [reply sensor/common/component.go]: Fixed

File: pkg/channelmultiplexer/channelmultiplexer.go
Comment (jschnath) [reply pkg/channelmultiplexer/channelmultiplexer.go]: 👍

I've also changed the multiplexer AddChannel to take `<-chan T`instead of `<-chan* T`

File: sensor/common/compliance/node_inventory_handler_impl.go
Comment (jschnath) [reply sensor/common/compliance/node_inventory_handler_impl.go]: Done

File: sensor/common/compliance/node_inventory_handler_impl.go
Comment (jschnath) [reply sensor/common/compliance/node_inventory_handler_impl.go]: Done and added a TODO for the followup regarding removing the Nack to have another look here

File: sensor/common/compliance/node_inventory_handler_impl.go
Comment (jschnath) [reply sensor/common/compliance/node_inventory_handler_impl.go]: Done

File: sensor/common/component.go
Comment (jschnath) [reply sensor/common/component.go]: Done

**Comment by vikin91** [comment]
@jschnath apparently the compilation is failing (https://github.com/stackrox/stackrox/actions/runs/4807324040/jobs/8555943465?pr=5165)
```
Compiling Go source in ./sensor/kubernetes to bin/linux_amd64/kubernetes
# github.com/stackrox/rox/sensor/common/compliance
sensor/common/compliance/multiplexer.go:75:18: cannot use channel (variable of type <-chan *common.MessageToComplianceWithAddress) as type <-chan common.MessageToComplianceWithAddress in argument to c.mp.AddChannel
sensor/common/compliance/multiplexer.go:80:9: cannot use c.mp.GetOutput() (value of type <-chan common.MessageToComplianceWithAddress) as type <-chan *common.MessageToComplianceWithAddress in return statement
make: *** [Makefile:411: main-build-nodeps] Error 2
Error: Process completed with exit code 2.
```

File: sensor/common/component.go
Comment (vikin91) [reply sensor/common/component.go]: https://issues.redhat.com/browse/ROX-16931

File: pkg/channelmultiplexer/channelmultiplexer_test.go
Comment (vikin91) [reply pkg/channelmultiplexer/channelmultiplexer_test.go]: I refactored it a bit additionally

**Review by vikin91 — APPROVED** [review]
This looks pretty good now! 
I will repeat the manual tests before merging/requesting more reviews

**Comment by jschnath** [comment]
Looks good, thanks for fixing it Piotr

**Pull request merged**
